cupy/cuda/jitify.pyx

```diff
@@ -143,15 +180,33 @@ cdef inline void init_cupy_headers() except*:
     _init_cupy_headers_from_scratch()
 
 
-init_cupy_headers()
+cpdef void _init_module() except*:
+    if _jitify_init:
+        return
+
+    global _jitify_cache_dir
+    if _jitify_cache_dir is None:
+        _jitify_cache_dir = os.getenv(
+            'CUPY_CACHE_DIR', os.path.expanduser('~/.cupy/jitify_cache'))
+
+    # Set up a version guard for invalidating the cache. Right now,
+    # we use the build-time versions of CUB/Jitify & CuPy version.
+    # TODO(leofang): Parse CUB/Thrust/libcu++ versions at process-
+    # start time, for enabling CCCL + CuPy developers?
+    global _jitify_cache_versions
+    if _jitify_cache_versions is None:
+        _jitify_cache_versions = (
+            f"{get_build_version()}_{cub.get_build_version()}_"
```

Comment:
Could you add NVRTC version as a cache key as well? I experienced an error when switching between CUDA 12 and CUDA 11 in my local.

Reply:
Hmmm, what's the error? We are caching headers here, not device code, so I am surprised that CUDA/NVRTC versions would matter.

Reply:
Sorry I couldn't reproduce the error now, but I think it's better to add the version as cache key because headers from CTK are added in the cache:

```
% jq '. | keys[]' ~/.cupy/jitify_cache/jitify_1a0ca0e_200200_2_a664cdd35aad76d128b93dfa58dbc1cf8876ffd2.json | grep 'cooperative_groups.h'
"../cooperative_groups.h"
"cooperative_groups.h"

% jq '. | keys[]' ~/.cupy/jitify_cache/jitify_1a0ca0e_200200_2_a664cdd35aad76d128b93dfa58dbc1cf8876ffd2.json | grep 'cuda_runtime_api.h'
"cuda_runtime_api.h"
```

Reply:
Ah yes, this remind me to I forgot to raise this discussion with you... I dislike caching CTK headers and think we should purge them (except for `cuda_fp16.h/hpp` which we vendor), see below. Thoughts?

Reply:
Ideally we wanna add cudart's version to the cache key, not nvrtc's, because these headers are coming from cudart (more precisely, cuda-cudart-dev). However, once we statically link to cudart, `runtimeGetVerson()` be no longer reflecting the actual (package or CTK) version seen at run time. This is tricky.

Reply:
Indeed it's tricky... I think adding NVRTC version is handy here. Another idea could be to calculate the hash of `cuda_runtime_api.h` (jitified code returned from Jitify) and add it to the cache file name. The header includes `CUDART_VERSION` constant.

Reply:
I pushed commit [306ec8d](https://github.com/cupy/cupy/pull/7934/commits/306ec8d28b0e39f9d6c0980bca6082357e0a1b97). I parse `CUDART_VERSION` from the header (if found). If it fails, I fall back to use NVRTC's version as the proxy.